### PR TITLE
Chainparams: Generic chainparam selection with -chain=<strNetworkID>

### PR DESCRIPTION
**Pull request by jtimon**
Continues #4804.
Deprecate -testnet and -regtest in favor of -network=testnet and -network=regtest.
When the old args are used a warning will be logged.
The repetition of the invalid combination error string is also avoided.
Additionally, the chainparam networkID is removed since it's not used anymore (as it shouldn't).

### CODE REVIEW

**Comment by jtimon** [comment]
Closing until #5598 gets merged.

**Comment by jtimon** [comment]
Reopened and rebased after #5598 has been merged. Now it's easier to review.

**Comment by petertodd** [comment]
Rather that having the argument be "-network" why don't we rename it to "-chain"?

Network is a term that's already overloaded in multiple contexts. After all, we called it *chain* params rather than network params.

**Comment by jtimon** [comment]
Yes, I also thought about "mode" and "chain". Technically main and unittest are the same chain, since they share the same genesis block and all, so I decided for "network". But that's the only exception so probably "chain" still makes sense.
Let's wait for other people to decide the color of the bike shed and I'll make the change to whatever is more popular.

**Comment by jtimon** [comment]
Needed rebase so I squashed the nits.

**Comment by laanwj** [comment]
I'm still not convinced that this is necessary. The reason for this change seems very academic; `-testnet` `-regtest` versus `-network=testnet` `-network=regtest`, meh.

**Comment by petertodd** [comment]
@laanwj FWIW I always use --testnet and --regtest in my command-line tools.

File: src/chainparamsbase.cpp [hunk 1]
```diff
@@ -94,22 +93,26 @@ CBaseChainParams::Network NetworkIdFromCommandLine()
...
+        throw std::runtime_error("Invalid combination of -regtest and -testnet.\n");
     if (fRegTest)
         return CBaseChainParams::REGTEST;
     if (fTestNet)
         return CBaseChainParams::TESTNET;
-    return CBaseChainParams::MAIN;
+
+    std::string chain = GetArg("-chain", "main");
+    if (chain == "main")
+        return CBaseChainParams::MAIN;
```
Comment (laanwj) [inline src/chainparamsbase.cpp]: What if these would be defined as string constants, e.g.:
```
    const std::string CBaseChainParams::MAIN = "main";
    const std::string CBaseChainParams::TESTNET = "testnet";
    const std::string CBaseChainParams::REGTEST = "regtest";
```
... I guess we could get rid of the `CBaseChainParams::Network` enumeration type completely, making `SelectParams` take a string. This kind of switching code would be no longer necessary apart from the data structure selection in SelectParams itself.

**Comment by laanwj** [comment]
This also needs changes to the qt autostart functionality, #5793, which relies on passing through -regtest/-testnet at the moment. But at least that is a simplification, you could just pass `-chain=%s` with %s the current chain name.

File: src/chainparamsbase.cpp (hunk 1 above)
Comment (jtimon) [inline src/chainparamsbase.cpp]: Well, we can get rid of the enumeration even if the strings aren't declared as constants. I won't opposed to it but working at big java projects were those type of constants are abundant I have always wondered if the gain in readability that CBaseChainParams::TESTNET provides compared to "testnet".
Yes, yes, developers could type "testtnet" and have some fun getting an "unkown chain 'testtnet'" exception...

Personally I like the Params(type) or Params(std::string) factory more than Params() in combination with SelectParams(). I mean, this is closer to what I would do: https://github.com/jtimon/bitcoin/commit/d135001a0a8ae9536553d49b1ac8a3c2e0643e41#diff-d22bc3e058f8982972e2eb381a1df668R34
In fact my preference when it comes to creating polymorphic objects would be for the factory function to return a pointer to a newly created object that the caller has to delete, a good-old factory without singleton [like Params()] or per-type singleton [like Params(std::string)], but that's more than we need. 
I was scared that moving from enum to string was going to be something people wouldn't like, but if that's not the case I can definitely improve this.
The remaining question is, do we really need the string constants when you will get an exception right in the face if you don't use the correct string?
In my opinion, "the same string you use in command line and it's documented" would be ideal.

**Comment by jtimon** [comment]
@laanwj could you give a line number link on where is that passed?

**Comment by jtimon** [comment]
@petertodd as discussed those options will remain usable and not deprecated.
Given that unittestnet has disappeared, it's really mainly an extensibility and altchain friendly patch so I understand if it remains in limbo until it is more directly useful for bitcoin.
It won't be hard to maintain unmerged anyway. Whenever somebody thinks about adding a third --harforktestnet or whatever this PR will be here to remind us that N is better than 2 + 1 (+ the implicit default "main" or "bitcoin").
We can label it low priority or I can put an uppercased word at the beginning of the tittle to clarify that people shouldn't get very distracted by this unless there's nothing else to merge or the somebody reanimates it [by proposing an additional chainparam option in the command line].
Paint it black and let it be, or something.
I can add a compilation error for a red cross to be shown.
But keeping it open I remember to rebase it and "police" this part of the code in some sense.
I can also just set an alarm to rebase it with the PR closed.
Not a strong opinion besides "if a chainparam option has to be added, it must be generic".
Sorry for the wall of text.

**Comment by jtimon** [comment]
Updated with @laanwj 's suggestion for replacing the enum with strings and adding a new GetParamsHelpMessages() function. Note that it could be used in src/bitcoin-cli.cpp too but I wasn't sure if we wanted that depending on ui_interface.

**Comment by jtimon** [comment]
Sorry, I forgot to deal with the qt autostart functionality as pointed out by@laanwj.
It is done now.

I also stopped using the bip70 nomenclature for the string constants since that changes "testnet" for just "test", which wouldn't have worked very well with the qt part.
Is there any possibility to correct bip70 in that sense?

File: src/qt/guiutil.cpp [hunk 2]
```diff
@@ -706,16 +704,15 @@ bool SetStartOnSystemStartup(bool fAutoStart)
...
-            optionFile << "Name=Bitcoin (testnet)\n";
-        else if (GetBoolArg("-regtest", false))
-            optionFile << "Name=Bitcoin (regtest)\n";
-        else
+        if (chain == "main")
             optionFile << "Name=Bitcoin\n";
-        optionFile << "Exec=" << pszExePath << strprintf(" -min -testnet=%d -regtest=%d\n", GetBoolArg("-testnet", false), GetBoolArg("-regtest", false));
+        else
+            optionFile << strprintf("Name=Bitcoin (%s)\n", chain);
+        optionFile << "Exec=" << pszExePath << strprintf(" -min -chain=%s -regtest=%d\n", chain);
```
Comment (laanwj) [inline src/qt/guiutil.cpp]: Spurious `-regtest=%d`

**Comment by jtimon** [comment]
Oopss, thanks @laanwj , fixed.
Also adding a commit in which src/bitcoin-cli.cpp also uses GetParamsHelpMessages() [and also ui_interface].
Thoughts on that commit? @theuni is there an specific reason for bitcoin-cli not to use ui_interface that I'm missing?

**Comment by jtimon** [comment]
I mean, the last commit would be kind of undoing https://github.com/bitcoin/bitcoin/commit/73ac7abd08a70adf22e24d0f5f7631e7f8b7c5bb

That's why I'm assuming people may not lie the idea.

**Comment by jtimon** [comment]
Never mind about the SQUASHME commits, I rebased on top of #6022, which is what I think makes most sense.

**Comment by jtimon** [comment]
Closing until #6022 is merged.

**Comment by jtimon** [comment]
Rebased after https://github.com/jtimon/bitcoin/commit/b74dcb3b4aa29958d22135f2d6c015578069e83c ( #6022  )

**Comment by jtimon** [comment]
Needed rebase

**Comment by jtimon** [comment]
Added 3 commits to hopefully make it more interesting.
The added tests are kind of redundant, but won't hurt.
I'm specially interested on @theuni 's opinion on https://github.com/jtimon/bitcoin/commit/65e2a722d64454378bb481afab154e4c9e1178d9 (since I know he wants to eventually separate the state-related functions out, but ParamsFactory() doesn't rely on any state).

**Comment by jtimon** [comment]
Removed https://github.com/jtimon/bitcoin/commit/ac1c555f275109582974ab0245a0d81854cf1ddd since is now a proposal for the trivial branch (https://github.com/theuni/bitcoin/pull/27).

To better read commit "Chainparams: The hash of the genesis block it's the genesis checkpoint and chain id", please rgrep hashGenesisBlock

All mentions of it are whether to:

- Its own asignments, of course.
- Make sure the genesis block is in a block index we're maintaining, like in ./init.cpp:1112, of course.
- Asserts in chainparams.cpp (no need to aditionally check genesis.hashMerkleRoot), let's keep this, why not?

- Avoid unnecessary checkpoints for the genesis block, in case we want to apply what's inside it somewhere, for example: ./main.cpp:1708, ./main.cpp:2641, ./main.cpp:2732, ./main.cpp:3385, ./main.cpp:3400, ./main.cpp:3495

Why the heck are we trying to check the genesis block in the first place?
The genesis block is not something to test, it is actually the first rule of the chain, it has to be right or the chain rules are broken (softforks and hardforks will come at a later nHeight).

But what about checkpoints?

The logic of checkpoints can always safely rely on the first checkpoint being true.
The last point may not be true if three's a reorg behind it, otherwise you're ready to optimize.
We could introduce a configurable third layer of "this is my new genesis no matter what", which would be good for pruning and skipping all validations on blocks that are too old to believe they can be reorg-ed, but this would only really make sense with some form of utxo commitment (did I say that append-only-txo + append-only-confirmed is enough for this?).

Therefore, we should not:

- Disperse hardcoded values.
- Double-check hashes of hashes
- Leave checkpoints without the only true first safe checkpoint: the genesis block

**Comment by jtimon** [comment]
Sorry, pushed too fast. Now I use containers instead of those ugly ClearSelectedParams functions, and the first commit actually compiles (now all of them do).

**Comment by jtimon** [comment]
Closing for now, I will open a version with some of the changes but without adding the new option.

**Comment by jtimon** [comment]
This is the promised version with some of the improvements but without adding the new option: #6229
Opened for discussion.

**Pull request closed without merging**